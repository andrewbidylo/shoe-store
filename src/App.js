import Card from './components/Card'
import Header from './components/Header'
import Drawer from './components/Drawer'
import React from 'react'
import axios from 'axios';



function App() {
  
  const [items, setItems] = React.useState([])
  const [cartOpened, setCartOpened] = React.useState(false)
  const [itemsForCard, setCartItem] = React.useState([])
  const [searchValue, serSearchValue] = React.useState('')

// Get items from server (first render) and hendeling errors.
// Set get data to state.

  React.useEffect(() => {
    axios.get('https://61c3afad9cfb8f0017a3ec85.mockapi.io/items')
      .then((res) => {
        setItems(res.data);
      })
      .catch((err) => {
        console.log(err);
      })
      axios.get('https://61c3afad9cfb8f0017a3ec85.mockapi.io/cart')
      .then((res) => {
        setCartItem(res.data);
      })
      .catch((err) => {
        console.log(err);
      })
  }, [])

  // Remove item from the Card. From the DB and DOM.
  const onRemoveItem = (id) => {
    axios.delete(`https://61c3afad9cfb8f0017a3ec85.mockapi.io/cart/${id}`)
        setCartItem(prev => prev.filter(item => item.id !== id));

  }

  // Add selected item to the DB and to the local state.
  const onAddtoCart = (obj) => {
    axios.post('https://61c3afad9cfb8f0017a3ec85.mockapi.io/cart', obj)
    setCartItem(prev => [...prev, obj])
  }


  return (
    <div className="wrapper clear">
      {cartOpened ? <Drawer onRemoveItem={onRemoveItem} itemsForCard={itemsForCard} onCloseDrawer={() => { setCartOpened(false) }} /> : null}
      <Header onClickCart={() => { setCartOpened(true) }} />
      <section className='content p-40'>
        <div className='d-flex align-center justify-between mb-40'>
          <h1>{searchValue ? `Search: ${searchValue}` : 'All Shoes'}</h1>
          <div className='search-block d-flex'>
            <img src='/img/search.svg' alt='Search' />
            {searchValue ? <img onClick={() => serSearchValue('')} className='clear remove-btn cu-p' src='/img/btn-remove.svg' alt='Clear' /> : null}
            <input onChange={(event) => serSearchValue(event.target.value)} value={searchValue} placeholder='Search'></input>
          </div>
        </div>
        <div className='allSneackers d-flex justify-between flex-wrap '>
          {items
            .filter((item) => item.title
              .toLowerCase()
              .includes(searchValue
                .toLowerCase()))
            .map((shoe, index) => (
              <Card
                key={index}
                title={shoe.title}
                price={shoe.price}
                imageURL={shoe.imageURL}
                onClickAdd={obj => { onAddtoCart(obj) }}
                onClickFavorite={() => { console.log('clicked favorite') }}
              />
            ))}


        </div>
      </section>
    </div>
  );
}

export default App;
