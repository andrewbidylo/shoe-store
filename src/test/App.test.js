import App from "../App"
import { screen, act, render } from '@testing-library/react';
import axios from 'axios';
import { MemoryRouter } from "react-router-dom";
import '@testing-library/jest-dom';
import ShallowRenderer from 'react-test-renderer/shallow';

// Shallow render
const renderer = new ShallowRenderer();

jest.mock('axios')

describe("App component", () => {

  // Setup App component. MemoryRouter keeps the history of your “URL” in memory 
  const setUp = props => render(
    <MemoryRouter >
      <App />
    </MemoryRouter>)

  afterEach(() => {
    jest.restoreAllMocks();
  });

  let response
  beforeEach(() => {
    axios.get.mockReturnValue(response)
    response = {
      data: [
        {
          "id": 1,
          "title": "Men's Boost ZK3 shoes 2022",
          "price": 122,
          "imageURL": "./img/shoes/1.png"
        },
        {
          "id": 2,
          "title": "Men's Grand Low shoes 2021",
          "price": 299,
          "imageURL": "./img/shoes/2.png"
        },
        {
          "id": 3,
          "title": "Women's Star Race shoes 2022",
          "price": 329,
          "imageURL": "./img/shoes/3.png"
        },
      ]
    }
  }
  )

  it('should render App component', async () => {
    const component = renderer.render(<App />)
    expect(component).toMatchSnapshot()
  })

  it("should call axios get 3 times", async () => {
    await act(async () => setUp())
    expect(axios.get).toBeCalledTimes(3)

  })
  it('should render item with the title = "Men`s Boost ZK3 shoes 2022"', async () => {
    await act(async () => setUp())
    const shoes = await screen.findByText("Men's Boost ZK3 shoes 2022")
    expect(shoes).toBeVisible()
  })

  // it('should show an error', async () => {
  //  const comp = render(<App />)
  //   expect(comp).toBeCalledWith("Some error!");
  // })
  
})


