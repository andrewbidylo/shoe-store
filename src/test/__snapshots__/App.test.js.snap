// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render App component 1`] = `
<Context.Provider
  value={
    Object {
      "favorites": Array [],
      "isItemAdded": [Function],
      "items": Array [],
      "itemsForCart": Array [],
      "onAddToFavorite": [Function],
      "onAddtoCart": [Function],
      "setCartItem": [Function],
      "setCartOpened": [Function],
    }
  }
>
  <div
    className="wrapper clear"
  >
    <Header
      onClickCart={[Function]}
    />
    <Routes>
      <Route
        element={
          <Home
            isLoading={true}
            items={Array []}
            itemsForCart={Array []}
            onAddToFavorite={[Function]}
            onAddtoCart={[Function]}
            searchValue=""
            setSearchValue={[Function]}
          />
        }
        exact={true}
        path="/"
      />
      <Route
        element={<Favorites />}
        exact={true}
        path="/favorites"
      />
      <Route
        element={<Orders />}
        exact={true}
        path="/orders"
      />
    </Routes>
  </div>
</Context.Provider>
`;
